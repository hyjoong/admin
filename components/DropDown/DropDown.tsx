import useOutsideClick from "@hooks/useOutsideClick";
import DownArrow from "@components/DropDown/DownArrow";
import React, { Dispatch, MouseEvent, SetStateAction, useState } from "react";
import styled, { css } from "styled-components";
import { CategoryToEnglish, periodToEnglish } from "@utils/transferLang";

interface Props {
  selectOption: string;
  exceptOption?: string;
  menuList: string[];
  order?: number;
  setItemSelect: Dispatch<SetStateAction<string>>;
}

const DropDown = ({
  selectOption,
  exceptOption,
  menuList,
  order,
  setItemSelect,
}: Props) => {
  const [isMenu, setIsMenu] = useState<boolean>(false);
  const handleMenu = () => {
    setIsMenu(!isMenu);
  };

  const handleClose = () => {
    setIsMenu(false);
  };

  const dorwpDownRef = useOutsideClick(handleClose);

  const handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    const selectValue = e.currentTarget.dataset.value;
    menuList.length === 2
      ? setItemSelect(periodToEnglish[selectValue])
      : setItemSelect(CategoryToEnglish[selectValue]);
    setIsMenu(false);
  };

  return (
    <DropDownContainer ref={dorwpDownRef}>
      <DropDownBox onClick={handleMenu} order={order}>
        {selectOption}
        <ArrowIcon isMenu={isMenu}>
          <DownArrow />
        </ArrowIcon>
      </DropDownBox>
      <OptionList>
        {isMenu &&
          menuList.map(
            (item, index) =>
              item !== exceptOption && (
                <Option key={index}>
                  <button data-value={item} onClick={(e) => handleClick(e)}>
                    {item}
                  </button>
                </Option>
              )
          )}
      </OptionList>
    </DropDownContainer>
  );
};

const DropDownContainer = styled.div`
  position: relative;

  &:not(:last-child) {
    margin-right: 10px;
  }
`;

const DropDownBox = styled.div<{ order: number }>`
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding: 10px 15px 10px 35px;
  width: 60px;
  height: 20px;
  border: 1px solid #d1d8dc;
  font-size: 14px;
  font-weight: 600;
  border-radius: 10px;
  cursor: pointer;

  ::before {
    position: absolute;
    width: 10px;
    height: 10px;
    top: 15px;
    left: 20px;
    content: "";
    border-radius: 50%;

    ${({ order }) =>
      order &&
      (order === 1
        ? css`
            background-color: #4fadf7;
          `
        : css`
            background-color: #85da47;
          `)};
  }
`;

const ArrowIcon = styled.div<{ isMenu: boolean }>`
  transition: transform 0.3s;
  ${({ isMenu }) =>
    isMenu &&
    css`
      transform: rotate(-180deg);
      transition: transform 0.3s;
    `}
`;

const OptionList = styled.ul`
  position: absolute;
  top: 45px;
  border-radius: 10px;
  border: 1px solid #d1d8dc;
  transition: all 0.2s ease-out;
  background: #ffffff;
  z-index: 2;
`;

const Option = styled.li`
  padding: 10px 15px;
  width: 90px;
  height: 20px;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;

  button {
    width: 100%;
    height: 100%;
    font-size: 14px;
    font-weight: 600;
  }

  &:not(:last-child) {
    border-bottom: 1px solid #d1d8dc;
  }
`;

export default DropDown;
