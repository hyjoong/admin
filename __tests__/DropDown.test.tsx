import DropDown from "@components/DropDown/DropDown";
import { render } from "@testing-library/react";

describe("DropDown", () => {
  const MOCKDROPDOWN = ["광고비", "노출 수", "클릭수", "전환 수", "매출"];
  const MockSelect = MOCKDROPDOWN[0];
  const setMockSelect = jest.fn();

  const renderDropDown = () => {
    return render(
      <DropDown
        selectOption={MockSelect}
        menuList={MOCKDROPDOWN}
        setItemSelect={setMockSelect}
      />
    );
  };

  it("DropDown selectOption이 화면에 보인다.", () => {
    const { container } = renderDropDown();
    expect(container).toHaveTextContent(MOCKDROPDOWN[0]);
    expect(container.firstChild).toMatchSnapshot();
  });

  //   isMenu가 false라서 제대로 동작 안 함
  //   it("DropDown 옵션들이 화면에 보인다.", () => {
  //     const { container } = renderDropDown();
  //     MOCKDROPDOWN.forEach((item) => {
  //       expect(container).toHaveTextContent(item);
  //     });
  //   });
});
