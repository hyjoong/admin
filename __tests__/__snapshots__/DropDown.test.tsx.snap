// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown DropDown selectOption이 화면에 보인다. 1`] = `
<div
  class="sc-bdnxRM knzmnr"
>
  <div
    class="sc-gtsrHT ddRnLq"
  >
    광고비
    <div
      class="sc-dlnjwi jzVlaj"
    >
      <svg
        fill="none"
        height="8"
        viewBox="0 0 12 8"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1.41 0.589966L6 5.16997L10.59 0.589966L12 1.99997L6 7.99997L0 1.99997L1.41 0.589966Z"
          fill="#3A474E"
        />
      </svg>
    </div>
  </div>
  <ul
    class="sc-hKFxyN dxCxFE"
  />
</div>
`;
